### PR TITLE
Finish gps support

### PR DESCRIPTION
**Pull request by tblaha**
To be able to generate identical output to bb-log-viewer also if GPS is included, I did the following:
- added a `last_gps_home_frame` variable to the `ctx` that is updated everytime GPS_HOME is parsed
- implement the predictors for GPS coordinates (latitude and longitude separately, which is caught explicitely in `reader.py`)
- ensured that only intra, slow and GPS fields are added to field_names (GPS home isn't written out by bb-log-viewer)
- add GPS data to the frame the same way slow frames have been
- updated `parser_test.py`. I ran 2 separate test cases through it, one with GPS and one without. However, more testing would be beneficial. @atomgomba would you still have some test `.csv` lying around?

### CODE REVIEW

**Comment by atomgomba** [comment]
@tblaha Once again, thank you for your contribution, I really appreciate it! :) Please rebase this onto master

**Comment by tblaha** [comment]
No problem!
I use this tool to read data into a pandas dataframe and do the fixed-point to floating-point unit conversions. Would you be interested to add this, for instance as a `.to_pandas(convert_units=True)` method, or something similar?

**Comment by atomgomba** [comment]
@tblaha Sorry for the delays, I have limited time, but I'm planning to test this today. If all goes well I'm going to push changes to the documentation, including README, and finally create a new PyPi release :)

File: orangebox/predictors.py [hunk 1]
```diff
@@ -64,6 +64,18 @@ def _increment(new: Number, ctx: Context) -> Number:
     return 1 + ctx.get_past_value(0) + ctx.count_skipped_frames()
 
 
+@map_to(7, predictor_map)
+def _home_coord_0(new: Number, ctx: Context) -> Number:
+    # TODO: store GPS frame history
```
Comment (atomgomba) [inline orangebox/predictors.py]: This TODO comment is now obsolete, please remove it

File: orangebox/predictors.py [hunk 2]
```diff
@@ -64,6 +64,18 @@ def _increment(new: Number, ctx: Context) -> Number:
...
+    return ctx.last_gps_home_frame.data[0] + new if ctx.last_gps_home_frame.data else 0
```
Comment (atomgomba) [inline orangebox/predictors.py]: Please put the second member of this addition in parentheses for readability and extract this long expression into a variable named `result`

Comment (atomgomba) [inline orangebox/predictors.py]: This returns an `int`, please divide `result` by `10_000_000` so it becomes directly usable by consumers

File: orangebox/predictors.py [hunk 3]
```diff
@@ -64,6 +64,18 @@ def _increment(new: Number, ctx: Context) -> Number:
...
+
+
+@map_to(256, predictor_map)
+def _home_coord_1(new: Number, ctx: Context) -> Number:
+    # TODO: store GPS frame history
```
Comment (atomgomba) [inline orangebox/predictors.py]: Obsolete comment, please remove

File: orangebox/predictors.py [hunk 4]
```diff
@@ -64,6 +64,18 @@ def _increment(new: Number, ctx: Context) -> Number:
...
+    return ctx.last_gps_home_frame.data[1] + new if ctx.last_gps_home_frame.data else 0
```
Comment (atomgomba) [inline orangebox/predictors.py]: Please extract this long expression as well and add parentheses

File: orangebox/reader.py [hunk 5]
```diff
@@ -140,6 +140,9 @@ def _build_field_defs(self):
                     field_defs[frame_type] = [FieldDef(frame_type) for _ in range(len(header_value))]
                 prop = header_key.split(" ", 2)[-1]
                 for i, framedef_value in enumerate(header_value):
+                    if (field_defs[frame_type][i].name == "GPS_coord[1]") \
```
Comment (atomgomba) [inline orangebox/reader.py]: Please remove the parentheses, it's not required in Python

Comment (atomgomba) [inline orangebox/reader.py]: Please extract the expression `field_defs[frame_type][i].name` into a variable called `fdef_name` and do not wrap the condition

File: orangebox/parser.py [hunk 6]
```diff
@@ -62,9 +62,13 @@ def set_log_index(self, index: int):
         self._headers = {k: v for k, v in reader.headers.items() if "Field" not in k}
         self._ctx = Context(self._headers, reader.field_defs)
         self._field_names = []
-        for fdef in reader.field_defs.values():
-            self._field_names += filter(lambda x: x is not None and x not in self._field_names,
-                                        map(lambda x: x.name, fdef))
+        for ftype in [FrameType.INTRA, FrameType.SLOW, FrameType.GPS]:
```
Comment (atomgomba) [inline orangebox/parser.py:65]: Nice catch! :+1:

File: orangebox/parser.py [hunk 7]
```diff
@@ -143,9 +161,27 @@ def frames(self) -> Iterator[Frame]:
                 ctx.invalid_frame_count += 1
                 continue
             last_iter = current_iter
-            if last_slow is not None:
-                # append data from previous SLOW frame
-                frame = Frame(ftype, frame.data + last_slow.data)
+
+            # add in extra frames (GPS, GPS_HOME and SLOW)
+            edata = []
```
Comment (atomgomba) [inline orangebox/parser.py]: Please give a more descriptive name to this variable, maybe `extra_data`?

File: parser_test.py [hunk 8]
```diff
@@ -13,14 +13,31 @@ def compare(frame_index, parser, frame, csv_frame, show_all_fields):
...
         csv_value = _trycast(csv_frame[i])
-        try:
-            fdef = reader.field_defs[frame.type][i]
-        except IndexError:
-            # account for the different numbers of field defs
-            fdeflen = len(reader.field_defs[frame.type])
-            fdef = reader.field_defs[FrameType.SLOW][i - fdeflen]
+
+        fdef = None
+        if (i < intra_inter_len):
```
Comment (atomgomba) [inline parser_test.py]: No parentheses please

File: parser_test.py [hunk 9]
```diff
@@ -13,14 +13,31 @@ def compare(frame_index, parser, frame, csv_frame, show_all_fields):
...
+            fdef = reader.field_defs[FrameType.INTER][i]
+        else:
+            curMaxLen = intra_inter_len
```
Comment (atomgomba) [inline parser_test.py]: Please use `snake_case` naming (everywhere in this file)

**Review by atomgomba — CHANGES_REQUESTED** [review]
The decoded output looks correct, I have made some remarks on the code

File: orangebox/predictors.py (hunk 4 above)
Comment (atomgomba) [inline orangebox/predictors.py]: Also divide the result by `10_000_000`

**Comment by atomgomba** [comment]
I just realized just blindly dividing by `10_000_000` would mess with the altitude value, if my thinking is right. Feel free to disregard this part, we can fix this once this is merged in a separate PR

**Comment by tblaha** [comment]
Thanks for the review, looks like I need to brush up on some PEP8! I addressed all your comments and ran the tests again.
I agree that processing variables (like converting units or converting fixed point to floating point) should probably be done in other methods. It's also not done by bb-log-viewer; it just dumps the raw values into the CSV for all other fields as well.

**Review by atomgomba — APPROVED** [review]

**Comment by atomgomba** [comment]
@tblaha Good job, thank you! :partying_face:

**Pull request merged**